sdk/tables/aztable/entity.go

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package aztable
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// https://docs.microsoft.com/en-us/rest/api/storageservices/payload-format-for-table-service-operations
+
+type Entity struct {
+	PartitionKey string
+	RowKey       string
+	Timestamp    EdmDateTime
+}
+
+type EdmEntity struct {
+	Metadata string `json:"odata.metadata"`
+	Id       string `json:"odata.id"`
+	EditLink string `json:"odata.editLink"`
+	Type     string `json:"odata.type"`
+	Etag     string `json:"odata.etag"`
+	Entity
+	Properties map[string]interface{} // Type assert the value to 1 of these: bool, int32, float64, string, EdmDateTime, EdmBinary, EdmGuid, EdmInt64
+}
+
+func (e EdmEntity) MarshalJSON() ([]byte, error) {
+	entity := map[string]interface{}{}
+	entity["PartitionKey"], entity["RowKey"] = e.PartitionKey, e.RowKey
+
+	for propName, propValue := range e.Properties {
+		entity[propName] = propValue
+		edmType := ""
+		switch propValue.(type) {
+		case EdmDateTime:
+			edmType = "Edm.DateTime"
+		case EdmBinary:
+			edmType = "Edm.Binary"
+		case EdmGuid:
+			edmType = "Edm.Guid"
+		case EdmInt64:
+			edmType = "Edm.Int64"
+		}
+		if edmType != "" {
+			entity[propName+"@odata.type"] = edmType
+		}
+	}
+	return json.Marshal(entity)
+}
+
+func (e *EdmEntity) UnmarshalJSON(data []byte) (err error) {
+	var entity map[string]json.RawMessage
+	err = json.Unmarshal(data, &entity)
+	if err != nil {
+		return
+	}
+	e.Properties = map[string]interface{}{}
+	for propName, propRawValue := range entity {
+		if strings.Contains(propName, "@odata.type") {
+			continue // Skip the @odata.type properties; we look them up explicitly later
+		}
+		switch propName {
+		// Look for EdmEntity's specific fields first
+		case "odata.metadata":
+			err = json.Unmarshal(propRawValue, &e.Metadata)
+		case "odata.id":
+			err = json.Unmarshal(propRawValue, &e.Id)
+		case "odata.editLink":
+			err = json.Unmarshal(propRawValue, &e.EditLink)
+		case "odata.type":
+			err = json.Unmarshal(propRawValue, &e.Type)
+		case "odata.etag":
+			err = json.Unmarshal(propRawValue, &e.Etag)
+		case "PartitionKey":
+			err = json.Unmarshal(propRawValue, &e.PartitionKey)
+		case "RowKey":
+			err = json.Unmarshal(propRawValue, &e.RowKey)
+		case "Timestamp":
+			err = json.Unmarshal(propRawValue, &e.Timestamp)
+		default:
+			// Try to find the EDM type for this property & get it's value
+			var propertyEdmTypeValue string = ""
+			if propertyEdmTypeRawValue, ok := entity[propName+"@odata.type"]; ok {
+				if err = json.Unmarshal(propertyEdmTypeRawValue, &propertyEdmTypeValue); err != nil {
+					return
+				}
+			}
+
+			var propValue interface{} = nil
+			switch propertyEdmTypeValue {
+			case "": // "<property>@odata.type" doesn't exist, infer the EDM type from the JSON type
+				// Try to unmarshal this property value as an int32 first
```

Comment:
I wonder if we should only try the int32 fallback if it deserialized as int64. My guess is that string will be more common.

Reply:
The service sends int64 values as strings; not as numbers. If the @odata.type:Edm.Int64 is present then we have to convert a string to an int64 integer. If the Edm.Int64 is missing that an int64 will be a string. A number will either be int32 or float64. In other words, I think this is right as is.

Reply:
My assumption was that this is to handle the default behavior of unmarshal, which makes all numbers Int64.

Reply:
json.Unmarshal converts all numbers to float64 by default.

Reply:
right - that is what I assuming this logic was meant to deal with, so that we don't up-convert int32 values to Int64 on unmarshal.